File: weird_captcha_gym/shared_runtime/app/mechanics/silent_colleague.js
(() => {
  "use strict";

  let model = null;
  let cleanup = null;
  const esc = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const mode = () => model.state.control_condition?.interaction || "full";

  function ticket() {
    const id = model.workshop.runtime_ticket_sequence[model.ticketIndex];
    return model.workshop.tickets.find((item) => item.id === id) || null;
  }

  function labelById(id) { return model.workshop.labels.find((item) => item.id === id) || null; }
  function fruitById(id) { return model.workshop.fruits.find((item) => item.id === id) || null; }
  function point(index) { return model.workshop.loop_points[index]; }

  function target() {
    const current = ticket();
    if (!current) return null;
    if (model.npcPhase === "signal") return labelById(current.label_id).station;
    return {
      jar: model.workshop.stations.jar_rack,
      handoff: model.workshop.stations.colleague_handoff,
      press: model.workshop.stations.colleague_press,
      hatch: model.workshop.stations.hatch,
    }[model.npcPhase] ?? null;
  }

  function moveNpc(direction) {
    const candidate = (model.npcPos + direction + model.workshop.loop_size) % model.workshop.loop_size;
    if (candidate === model.playerPos) {
      model.jams += 1;
      model.bump = "colleague";
      model.message = "LOOP BLOCKED";
    } else {
      model.npcPos = candidate;
      model.bump = null;
    }
  }

  function npcStep() {
    if (model.terminalFailure || !ticket()) return;
    const current = ticket();
    if (model.npcPhase === "press_wait") {
      if (model.tick > model.primeUntil) {
        model.spoils += 1;
        model.npcPhase = "signal";
        model.npcWait = 0;
        model.primeUntil = null;
        model.message = `SEAL MISSED · ${model.spoils}/${model.parameters.max_spoils} JARS SPOILED`;
        if (model.spoils >= model.parameters.max_spoils) {
          model.terminalFailure = true;
          model.message = "SHIFT FAILED · TOO MANY JARS SPOILED";
        }
      }
      return;
    }
    const destination = target();
    if (destination == null) return;
    if (model.npcPos !== destination) {
      moveNpc(Number(current.direction));
      if (model.npcPhase === "signal" && model.npcPos === destination) {
        model.npcWait = 1;
        model.message = "LABEL RACK VISITED";
      }
      return;
    }
    if (model.npcPhase === "signal") {
      if (model.npcWait <= 0) {
        model.npcWait = 1;
        model.message = "LABEL RACK VISITED";
      } else if (model.npcWait < model.parameters.signal_ticks) {
        model.npcWait += 1;
        model.message = "LABEL RACK VISITED";
      } else {
        model.npcWait = 0;
        model.npcPhase = "jar";
        model.message = "LABEL RACK CLEARED";
        moveNpc(Number(current.direction));
      }
    } else if (model.npcPhase === "jar") {
      model.npcPhase = "handoff";
      model.message = "EMPTY JAR LIFTED";
    } else if (model.npcPhase === "handoff") {
      if (model.shelf === current.fruit_id) {
        model.shelf = null;
        model.npcPhase = "press";
        model.message = "FRUIT TAKEN FROM PASS";
      } else if (model.shelf) {
        model.message = "PASS REFUSED";
        model.sigh = true;
      } else {
        model.message = "PASS EMPTY";
      }
    } else if (model.npcPhase === "press") {
      model.npcPhase = "press_wait";
      model.primeUntil = model.tick + model.parameters.press_window_ticks - 1;
      model.message = "HANDLE FLASH";
    } else if (model.npcPhase === "hatch") {
      model.delivered.push(current.id);
      model.ticketIndex += 1;
      model.npcPhase = "signal";
      model.npcWait = 0;
      model.message = ticket() ? "JAR DELIVERED" : "EVERY TICKET FILLED";
    }
  }

  function advance() {
    if (!model || model.submitting || model.completed) return;
    model.tick += 1;
    model.sigh = false;
    npcStep();
    if (!ticket() && !model.terminalFailure) model.ready = true;
  }

  function advanceToNow() {
    if (!model || model.submitting || model.completed) return;
    const target = Math.floor((performance.now() - model.started) / Number(model.parameters.tick_ms) + 1e-7);
    while (model.tick < target) advance();
  }

  function stationFruit() {
    return model.workshop.fruits.find((item) => Number(item.station) === model.playerPos)?.id || null;
  }

  function applyAction(action) {
    const before = model.playerPos;
    if (action === "ccw" || action === "cw") {
      const direction = action === "ccw" ? -1 : 1;
      const candidate = (before + direction + model.workshop.loop_size) % model.workshop.loop_size;
      const moved = candidate !== model.npcPos;
      if (moved) {
        model.playerPos = candidate;
        model.bump = null;
      } else {
        model.jams += 1;
        model.bump = "player";
        model.message = "OCCUPIED TILE";
      }
      return {kind: "move", action, from: before, to: model.playerPos, moved};
    }
    let effect = "idle";
    const fruit = stationFruit();
    const stations = model.workshop.stations;
    if (fruit) {
      model.playerCarrying = fruit;
      effect = "pick_fruit";
      model.message = `${fruitById(fruit).name} CRATE OPENED · CARRYING ONE BATCH`;
    } else if (model.playerPos === stations.handoff) {
      if (model.playerCarrying) {
        model.shelf = model.playerCarrying;
        model.playerCarrying = null;
        effect = "place_handoff";
        model.message = "FRUIT LEFT ON THE SHARED PASS";
      } else if (model.shelf) {
        model.playerCarrying = model.shelf;
        model.shelf = null;
        effect = "retrieve_handoff";
        model.message = "FRUIT RETRIEVED FROM THE PASS";
      }
    } else if (model.playerPos === stations.player_press && model.npcPhase === "press_wait" && model.primeUntil != null && model.tick <= model.primeUntil) {
      model.npcPhase = "hatch";
      model.primeUntil = null;
      effect = "paired_press";
      model.message = "BOTH HANDLES LANDED · JAR SEALED";
    } else {
      model.message = "NO AVAILABLE ACTION AT THIS TILE";
    }
    return {kind: "use", action: "use", position: model.playerPos, effect, carrying: model.playerCarrying, shelf: model.shelf};
  }

  function act(action, inputSource) {
    advanceToNow();
    if (!model || model.submitting || model.completed || model.terminalFailure) return;
    const claim = applyAction(action);
    model.events.push({sequence: model.events.length + 1, tick: model.tick, input_source: inputSource, ...claim});
    renderSurface();
  }

  function visibleIntent() {
    const current = ticket();
    if (!current) return "";
    const atSignal = model.npcPhase === "signal" && model.npcPos === labelById(current.label_id).station;
    if (model.parameters.intent_mode === "fruit_badge") {
      const fruit = fruitById(current.fruit_id);
      return `<span class="sc-intent" style="--intent:${fruit.hue}">${esc(fruit.glyph)}</span>`;
    }
    if (model.parameters.intent_mode === "label_badge" || (model.parameters.intent_mode === "hover_badge" && atSignal)) {
      const label = labelById(current.label_id);
      return `<span class="sc-intent" style="--intent:${label.hue}">${esc(label.sigil)}</span>`;
    }
    return "";
  }

  function stationMarkup() {
    const stations = [];
    model.workshop.fruits.forEach((fruit) => stations.push({index: fruit.station, cls: "fruit", title: fruit.name, glyph: fruit.glyph, hue: fruit.hue}));
    model.workshop.labels.forEach((label) => stations.push({index: label.station, cls: "label", title: `${label.name} LABEL`, glyph: label.sigil, hue: label.hue}));
    const fixed = model.workshop.stations;
    stations.push(
      {index: fixed.handoff, cls: "handoff player-handoff", title: "YOUR PASS", glyph: model.shelf ? fruitById(model.shelf).glyph : "⇄", hue: model.shelf ? fruitById(model.shelf).hue : "#d6c49c"},
      {index: fixed.player_press, cls: "press player-press", title: "YOUR HANDLE", glyph: "Ⅰ", hue: "#e8a84d"},
      {index: fixed.colleague_press, cls: "press colleague-press", title: "THEIR HANDLE", glyph: "Ⅱ", hue: "#6fbfc1"},
      {index: fixed.colleague_handoff, cls: "handoff colleague-handoff", title: "THEIR PASS", glyph: model.shelf ? fruitById(model.shelf).glyph : "⇄", hue: model.shelf ? fruitById(model.shelf).hue : "#d6c49c"},
      {index: fixed.jar_rack, cls: "jar", title: "JAR RACK", glyph: "▱", hue: "#d6e5dc"},
      {index: fixed.hatch, cls: "hatch", title: "DELIVERY HATCH", glyph: "↥", hue: "#b8d692"},
    );
    return stations.map((item) => {
      const [x, y] = point(item.index);
      const activePress = item.cls.includes("player-press") && model.npcPhase === "press_wait";
      return `<div class="sc-station ${item.cls} ${activePress ? "is-primed" : ""}" style="--x:${x}%;--y:${y}%;--station:${item.hue}" data-station-index="${item.index}"><b>${esc(item.glyph)}</b><span>${esc(item.title)}</span></div>`;
    }).join("");
  }

  function avatarMarkup(role, index) {
    const [x, y] = point(index);
    const carrying = role === "player" ? model.playerCarrying : (model.npcPhase === "jar" || model.npcPhase === "handoff" || model.npcPhase === "press" || model.npcPhase === "press_wait" ? "jar" : null);
    const bump = model.bump === role ? "is-bumped" : "";
    return `<div class="sc-avatar ${role} ${bump} ${role === "colleague" && model.sigh ? "is-sighing" : ""}" style="--x:${x}%;--y:${y}%" data-loop-index="${index}">
      ${role === "colleague" ? visibleIntent() : ""}<i><b>${role === "player" ? "YOU" : "Ⅱ"}</b></i>
      ${carrying ? `<em>${carrying === "jar" ? "▱" : esc(fruitById(carrying).glyph)}</em>` : ""}
      <span>${role === "player" ? "PRESERVER" : "COLLEAGUE"}</span>
    </div>`;
  }

  function ticketMarkup(item) {
    const delivered = model.delivered.includes(item.id);
    const label = labelById(item.label_id);
    const fruit = fruitById(item.fruit_id);
    return `<article class="sc-ticket ${delivered ? "is-done" : ""}">
      <div style="--label:${label.hue}">${esc(label.sigil)}</div><span><small>LABEL ${esc(label.name)}</small><strong>${esc(fruit.glyph)} · ${esc(fruit.name)}</strong></span><b>${delivered ? "SEALED" : "OPEN"}</b>
    </article>`;
  }

  function snapshot() {
    return {
      tick: model.tick, player_pos: model.playerPos, npc_pos: model.npcPos, player_carrying: model.playerCarrying,
      shelf: model.shelf, ticket_index: model.ticketIndex, npc_phase: model.npcPhase, npc_wait: model.npcWait,
      prime_until: model.primeUntil, delivered: [...model.delivered], spoils: model.spoils, jams: model.jams,
      terminal_failure: model.terminalFailure,
    };
  }

  function bindControls() {
    document.querySelector('[data-sc-act="ccw"]')?.addEventListener("click", () => act("ccw", "proxy_step"));
    document.querySelector('[data-sc-act="cw"]')?.addEventListener("click", () => act("cw", "proxy_step"));
    document.querySelector('[data-sc-act="use"]')?.addEventListener("click", () => act("use", "proxy_action"));
    document.getElementById("sc-certify")?.addEventListener("click", submit);
  }

  function renderSurface() {
    const root = document.querySelector(".silent-colleague");
    if (!root || !model) return;
    root.querySelector(".sc-floor").innerHTML = `<div class="sc-counter"><small>SHARED LINE</small><b>QUIET PRESERVES</b><span>ONE TILE · TWO WORKERS</span></div>${stationMarkup()}${avatarMarkup("player", model.playerPos)}${avatarMarkup("colleague", model.npcPos)}`;
    root.querySelector(".sc-ticket-list").innerHTML = model.workshop.tickets.map(ticketMarkup).join("");
    const pressRemaining = model.npcPhase === "press_wait" ? Math.max(0, model.primeUntil - model.tick + 1) : 0;
    root.querySelector(".sc-telemetry").innerHTML = `<div><small>WORK LINE</small><b>${model.ready ? "SHIFT COMPLETE" : model.terminalFailure ? "SHIFT STOPPED" : "IN MOTION"}</b></div><div><small>PRESS WINDOW</small><b class="${pressRemaining ? "is-hot" : ""}">${pressRemaining ? `${pressRemaining} TICKS` : "DORMANT"}</b></div><div><small>SPOILED</small><b>${model.spoils}/${model.parameters.max_spoils}</b></div><div><small>LOOP BLOCKS</small><b>${model.jams}</b></div>`;
    root.querySelector(".sc-message").textContent = model.message;
    const proxy = root.querySelector(".sc-proxy");
    if (proxy) proxy.hidden = mode() !== "simplified";
    const certify = document.getElementById("sc-certify");
    certify.disabled = model.submitting || model.completed;
    certify.textContent = model.ready ? "CERTIFY FILLED SHIFT" : model.terminalFailure ? "CERTIFY FAILED SHIFT" : "END & CERTIFY SHIFT";
    root.dataset.phase = model.npcPhase;
    root.dataset.ready = model.ready ? "true" : "false";
  }

  async function submit() {
    advanceToNow();
    if (!model || model.submitting || model.completed) return;
    const current = model;
    current.submitting = true;
    current.message = "INDEPENDENTLY REPLAYING EVERY STEP AND COLLEAGUE TICK…";
    renderSurface();
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify({
        mechanic_id: current.state.mechanic_id, task_id: current.state.task_id, challenge_id: current.state.challenge_id,
        interaction_mode: mode(), events: current.events, final_tick: current.tick, final_state: snapshot(), completed: current.ready,
      })});
      const outcome = await response.json();
      if (outcome.passed === true) {
        current.completed = true;
        current.message = "PASS · ALL TICKETS AND JOINT PRESSES REPLAYED";
        current.helpers.setReadout("PASS", "passed");
        renderSurface();
        document.querySelector(".silent-colleague")?.insertAdjacentHTML("beforeend", '<div class="sc-verdict is-pass"><small>SHIFT LEDGER ACCEPTED</small><strong>PASS</strong><span>THE LINE HELD TOGETHER</span></div>');
      } else if (outcome.passed === false && outcome.state) {
        const helpers = current.helpers;
        await render(outcome.state, helpers, {freshFailure: true});
        model.helpers.setReadout("FAIL", "error");
      } else {
        current.submitting = false;
        current.message = "CERTIFICATION REJECTED";
        current.helpers.setReadout("FAIL", "error");
        renderSurface();
      }
    } catch (_error) {
      if (model === current) {
        current.submitting = false;
        current.message = "CERTIFICATION LINK OFFLINE";
        current.helpers.setReadout("FAIL", "error");
        renderSurface();
      }
    }
  }

  function onKey(event) {
    if (event.repeat || mode() !== "full") return;
    if (["ArrowLeft", "ArrowRight", "a", "A", "d", "D", " "].includes(event.key)) event.preventDefault();
    if (event.key === "ArrowLeft" || event.key === "a" || event.key === "A") act("ccw", "keyboard_move");
    else if (event.key === "ArrowRight" || event.key === "d" || event.key === "D") act("cw", "keyboard_move");
    else if (event.key === " ") act("use", "keyboard_action");
  }

  async function render(state, helpers, options = {}) {
    cleanup?.();
    document.body.dataset.mechanic = "silent-colleague";
    model = {
      state, helpers, workshop: JSON.parse(JSON.stringify(state.workshop)), parameters: JSON.parse(JSON.stringify(state.parameters)),
      tick: 0, started: performance.now(), playerPos: state.workshop.player_start, npcPos: state.workshop.colleague_start,
      playerCarrying: null, shelf: null, ticketIndex: 0, npcPhase: "signal", npcWait: 0, primeUntil: null,
      delivered: [], spoils: 0, jams: 0, terminalFailure: false, ready: false, completed: false, submitting: false,
      bump: null, sigh: false, events: [], message: options.freshFailure ? "FAIL · FRESH SHIFT LOADED · LEDGER CLEARED" : "SHIFT OPEN",
    };
    helpers.app.innerHTML = `<section class="silent-colleague mode-${mode()}" data-mechanic="${esc(state.mechanic_id)}" data-interaction="${mode()}" data-fresh-failure="${options.freshFailure ? "true" : "false"}">
      <header class="sc-masthead"><div><small>BRAMBLE & BRINE CO-OPERATIVE · ${esc(state.workshop.batch_code)}</small><h1>${esc(state.prompt)}</h1></div><div class="sc-shift"><i></i><span>${mode().toUpperCase()} INPUT</span><b>D${state.control_condition?.difficulty || 4} · TICK ${state.parameters.tick_ms}MS</b></div></header>
      <main><section class="sc-workshop"><div class="sc-floor" aria-label="one-tile preservation workshop loop"></div><div class="sc-message readout" data-status="idle"></div></section><aside class="sc-orders"><header><small>DELIVERY HATCH / SHIFT LEDGER</small><h2>PRESERVE TICKETS</h2></header><div class="sc-ticket-list"></div><div class="sc-key"><span><i class="you"></i>YOU</span><span><i class="them"></i>COLLEAGUE</span></div></aside></main>
      <footer><div class="sc-telemetry"></div><div class="sc-proxy" hidden><button data-sc-act="ccw">↶ STEP</button><button data-sc-act="use">USE STATION</button><button data-sc-act="cw">STEP ↷</button></div><div class="sc-full-help">A / ← COUNTER-CLOCKWISE &nbsp; · &nbsp; D / → CLOCKWISE &nbsp; · &nbsp; SPACE USE</div><button id="sc-certify">END & CERTIFY SHIFT</button></footer>
      ${options.freshFailure ? '<div class="sc-verdict is-fail"><small>PRIOR LEDGER REJECTED</small><strong>FAIL</strong><span>FRESH SHIFT ISSUED</span></div>' : ""}
      ${helpers.cheatPanelTemplate()}
    </section>`;
    renderSurface();
    bindControls();
    helpers.installCheatPanel();
    window.addEventListener("keydown", onKey);
    let timer;
    const frame = () => {
      const before = model.tick;
      advanceToNow();
      if (model.tick !== before) renderSurface();
      timer = requestAnimationFrame(frame);
    };
    timer = requestAnimationFrame(frame);
    cleanup = () => { cancelAnimationFrame(timer); window.removeEventListener("keydown", onKey); };
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.silent_colleague = {rootSelector: ".silent-colleague", render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/silent_colleague.py
from __future__ import annotations

import copy
from typing import Any


MECHANIC_ID = "silent_colleague"


def _fail(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": message}


def _ticket(sim: dict[str, Any]) -> dict[str, Any] | None:
    sequence = sim["workshop"]["runtime_ticket_sequence"]
    if sim["ticket_index"] >= len(sequence):
        return None
    ticket_id = sequence[sim["ticket_index"]]
    return next(item for item in sim["workshop"]["tickets"] if item["id"] == ticket_id)


def _label(sim: dict[str, Any], label_id: str) -> dict[str, Any]:
    return next(item for item in sim["workshop"]["labels"] if item["id"] == label_id)


def _target(sim: dict[str, Any]) -> int | None:
    phase = sim["npc_phase"]
    ticket = _ticket(sim)
    if ticket is None:
        return None
    if phase == "signal":
        return int(_label(sim, ticket["label_id"])["station"])
    return {
        "jar": sim["workshop"]["stations"]["jar_rack"],
        "handoff": sim["workshop"]["stations"]["colleague_handoff"],
        "press": sim["workshop"]["stations"]["colleague_press"],
        "hatch": sim["workshop"]["stations"]["hatch"],
    }.get(phase)


def _initial(public: dict[str, Any]) -> dict[str, Any]:
    workshop = copy.deepcopy(public["workshop"])
    return {
        "workshop": workshop, "parameters": copy.deepcopy(public["parameters"]),
        "tick": 0, "player_pos": int(workshop["player_start"]), "npc_pos": int(workshop["colleague_start"]),
        "player_carrying": None, "shelf": None, "ticket_index": 0, "npc_phase": "signal",
        "npc_wait": 0, "prime_until": None, "delivered": [], "spoils": 0, "jams": 0,
        "terminal_failure": False,
    }


def _move_npc(sim: dict[str, Any], target: int, direction: int) -> None:
    size = int(sim["workshop"]["loop_size"])
    candidate = (int(sim["npc_pos"]) + direction) % size
    if candidate == int(sim["player_pos"]):
        sim["jams"] += 1
    else:
        sim["npc_pos"] = candidate


def _npc_step(sim: dict[str, Any]) -> None:
    if sim["terminal_failure"] or _ticket(sim) is None:
        return
    ticket = _ticket(sim)
    phase = sim["npc_phase"]
    if phase == "press_wait":
        if sim["tick"] > int(sim["prime_until"]):
            sim["spoils"] += 1
            sim["npc_phase"] = "signal"
            sim["npc_wait"] = 0
            sim["prime_until"] = None
            if sim["spoils"] >= int(sim["parameters"]["max_spoils"]):
                sim["terminal_failure"] = True
        return
    target = _target(sim)
    if target is None:
        return
    if int(sim["npc_pos"]) != int(target):
        _move_npc(sim, int(target), int(ticket["direction"]))
        if sim["npc_phase"] == "signal" and int(sim["npc_pos"]) == int(target):
            sim["npc_wait"] = 1
        return
    if phase == "signal":
        if sim["npc_wait"] <= 0:
            sim["npc_wait"] = 1
        elif sim["npc_wait"] < int(sim["parameters"]["signal_ticks"]):
            sim["npc_wait"] += 1
        else:
            sim["npc_wait"] = 0
            sim["npc_phase"] = "jar"
            _move_npc(sim, int(_target(sim)), int(ticket["direction"]))
    elif phase == "jar":
        sim["npc_phase"] = "handoff"
    elif phase == "handoff":
        if sim["shelf"] == ticket["fruit_id"]:
            sim["shelf"] = None
            sim["npc_phase"] = "press"
    elif phase == "press":
        sim["npc_phase"] = "press_wait"
        sim["prime_until"] = sim["tick"] + int(sim["parameters"]["press_window_ticks"]) - 1
    elif phase == "hatch":
        sim["delivered"].append(ticket["id"])
        sim["ticket_index"] += 1
        sim["npc_phase"] = "signal"
        sim["npc_wait"] = 0


def advance_to(sim: dict[str, Any], tick: int) -> None:
    if tick < sim["tick"]:
        raise ValueError("event tick moved backward")
    while sim["tick"] < tick:
        sim["tick"] += 1
        _npc_step(sim)


def _station_fruit(sim: dict[str, Any]) -> str | None:
    for fruit in sim["workshop"]["fruits"]:
        if int(fruit["station"]) == int(sim["player_pos"]):
            return str(fruit["id"])
    return None


def apply_action(sim: dict[str, Any], action: str) -> dict[str, Any]:
    before = int(sim["player_pos"])
    if action in {"ccw", "cw"}:
        direction = -1 if action == "ccw" else 1
        candidate = (before + direction) % int(sim["workshop"]["loop_size"])
        moved = candidate != int(sim["npc_pos"])
        if moved:
            sim["player_pos"] = candidate
        else:
            sim["jams"] += 1
        return {"kind": "move", "action": action, "from": before, "to": int(sim["player_pos"]), "moved": moved}
    if action != "use":
        raise ValueError("unknown player action")
    effect = "idle"
    fruit = _station_fruit(sim)
    stations = sim["workshop"]["stations"]
    if fruit is not None:
        sim["player_carrying"] = fruit
        effect = "pick_fruit"
    elif int(sim["player_pos"]) == int(stations["handoff"]):
        if sim["player_carrying"] is not None:
            sim["shelf"] = sim["player_carrying"]
            sim["player_carrying"] = None
            effect = "place_handoff"
        elif sim["shelf"] is not None:
            sim["player_carrying"] = sim["shelf"]
            sim["shelf"] = None
            effect = "retrieve_handoff"
    elif int(sim["player_pos"]) == int(stations["player_press"]) and sim["npc_phase"] == "press_wait" and sim["prime_until"] is not None and sim["tick"] <= int(sim["prime_until"]):
        sim["npc_phase"] = "hatch"
        sim["prime_until"] = None
        effect = "paired_press"
    return {"kind": "use", "action": "use", "position": int(sim["player_pos"]), "effect": effect, "carrying": sim["player_carrying"], "shelf": sim["shelf"]}


def snapshot(sim: dict[str, Any]) -> dict[str, Any]:
    return {
        "tick": int(sim["tick"]), "player_pos": int(sim["player_pos"]), "npc_pos": int(sim["npc_pos"]),
        "player_carrying": sim["player_carrying"], "shelf": sim["shelf"], "ticket_index": int(sim["ticket_index"]),
        "npc_phase": sim["npc_phase"], "npc_wait": int(sim["npc_wait"]), "prime_until": sim["prime_until"],
        "delivered": list(sim["delivered"]), "spoils": int(sim["spoils"]), "jams": int(sim["jams"]),
        "terminal_failure": bool(sim["terminal_failure"]),
    }


def _contract(truth: dict[str, Any], public: dict[str, Any]) -> tuple[dict[str, Any], str]:
    if truth.get("workshop") != public.get("workshop") or truth.get("parameters") != public.get("parameters"):
        raise ValueError("public workshop differs from grading truth")
    condition = truth.get("control_condition")
    if condition != public.get("control_condition"):
        raise ValueError("control condition mismatch")
    if condition is not None and condition.get("difficulty_parameters") != truth.get("parameters"):
        raise ValueError("difficulty parameters are not bound")
    interaction = str((condition or {}).get("interaction") or "full")
    if interaction not in {"simplified", "full"}:
        raise ValueError("invalid interaction mode")
    workshop = truth.get("workshop")
    if not isinstance(workshop, dict) or len(workshop.get("runtime_ticket_sequence") or []) != int(truth["parameters"]["ticket_count"]):
        raise ValueError("ticket sequence is incomplete")
    ticket_ids = {item.get("id") for item in workshop.get("tickets") or []}
    if set(workshop["runtime_ticket_sequence"]) != ticket_ids:
        raise ValueError("ticket inventory does not match colleague schedule")
    return workshop, interaction


def grade(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    for item in (payload, truth, public):
        if str(item.get("mechanic_id") or "") != MECHANIC_ID:
            return _fail("mechanic mismatch")
    for key in ("task_id", "challenge_id"):
        expected = str(truth.get(key) or "")
        if not expected or str(public.get(key) or "") != expected or str(payload.get(key) or "") != expected:
            return _fail(f"stale or mismatched {key}")
    try:
        _workshop, interaction = _contract(truth, public)
    except (KeyError, TypeError, ValueError) as exc:
        return _fail(f"invalid colleague contract: {exc}")
    if payload.get("interaction_mode") != interaction:
        return _fail("submitted interaction mode differs from task condition")
    events = payload.get("events")
    final_tick = payload.get("final_tick")
    if not isinstance(events, list) or not events:
        return _fail("player action transcript is empty")
    if isinstance(final_tick, bool) or not isinstance(final_tick, int) or final_tick < 0 or final_tick > 20000:
        return _fail("final tick is invalid")
    sources = {"full": {"move": "keyboard_move", "use": "keyboard_action"}, "simplified": {"move": "proxy_step", "use": "proxy_action"}}
    sim = _initial(public)
    try:
        for sequence, event in enumerate(events, 1):
            if not isinstance(event, dict) or event.get("sequence") != sequence:
                raise ValueError(f"event {sequence} has invalid sequence")
            tick = event.get("tick")
            if isinstance(tick, bool) or not isinstance(tick, int) or tick > final_tick:
                raise ValueError(f"event {sequence} has invalid tick")
            advance_to(sim, tick)
            expected = apply_action(sim, str(event.get("action") or ""))
            if event.get("input_source") != sources[interaction][expected["kind"]]:
                raise ValueError(f"event {sequence} uses the wrong input surface")
            claim = {key: value for key, value in event.items() if key not in {"sequence", "tick", "input_source"}}
            if claim != expected:
                raise ValueError(f"event {sequence} action claim disagrees with replay")
        advance_to(sim, final_tick)
    except (KeyError, StopIteration, TypeError, ValueError) as exc:
        return _fail(f"colleague replay rejected: {exc}")
    expected_final = snapshot(sim)
    if payload.get("final_state") != expected_final:
        return _fail("submitted final workshop state disagrees with replay")
    all_ids = list(public["workshop"]["runtime_ticket_sequence"])
    passed = expected_final["delivered"] == all_ids and not expected_final["terminal_failure"] and payload.get("completed") is True
    return {
        "graded": True,
        "passed": passed,
        "feedback": f"colleague replay delivered {len(expected_final['delivered'])}/{len(all_ids)} tickets; {expected_final['spoils']} spoiled; {expected_final['jams']} occupied-loop blocks",
    }
